metagraph/src/graph/alignment/aligner_seeder_methods.cpp

```diff
@@ -81,57 +84,234 @@ void ExactSeeder<NodeType>::call_seeds(std::function<void(Seed&&)> callback) con
+    if (canonical) {
+        // find sub-k matches in the reverse complement
+        std::string query_rc(this->query_);
+        reverse_complement(query_rc.begin(), query_rc.end());
+
+        // matching is done query prefix -> node suffix, so the query index of
+        // a match to the reverse complement is not known beforehand
+        // e.g.,
+        // k = 6;
+        // rev: rev_end_pos = 8
+        //     j
+        //     ****--      <-- start position in forward depends on match length
+        // GCTAGCATCTGAGAGGGGA fwd
+        // TCCCCTCTCAGATGCTAGC rc
+        //          --****
+        //            i    <-- match returned from call
+        for (size_t i = 0; i + this->config_.min_seed_length <= query_rc.size(); ++i) {
+            // initial estimate of the max seed length
+            size_t max_seed_length = std::min({ this->config_.max_seed_length,
+                                                this->graph_.get_k() - 1,
+                                                this->query_.size() - i });
 
-            if (!found) {
-                last_seed_size = std::max(last_seed_size - 1,
-                                          this->config_.min_seed_length);
+            // the reverse complement of the sub-k match will fall somewhere in this range
+            size_t j_min = query_rc.size() - i - max_seed_length;
+            size_t j_max = query_rc.size() - i - this->config_.min_seed_length;
+
+            // skip over positions which have better matches
+            while (min_seed_length[j_min] > max_seed_length && j_min <= j_max) {
+                ++j_min;
+                --max_seed_length;
             }
+
+            if (j_min > j_max)
+                continue;
+
+            const auto &boss = dbg_succ_.get_boss();
+
+            auto encoded = boss.encode({ query_rc.data() + i, max_seed_length });
+            auto [first, last, end] = boss.index_range(encoded.begin(), encoded.end());
+
+            size_t seed_length = end - encoded.begin();
+            size_t j = query_rc.size() - i - seed_length;
+
+            assert(seed_length < this->config_.min_seed_length
+                || j < min_seed_length.size());
+
+            if (seed_length < this->config_.min_seed_length
+                    || seed_length < min_seed_length[j]) {
+                continue;
+            }
+
+            min_seed_length[j] = seed_length;
+
+            // clear out shorter seeds
+            size_t s = seed_length;
+            for (size_t m = j + 1;
+                    m < min_seed_length.size() && s > min_seed_length[m];
+                    ++m) {
+                assert(suffix_seeds[m].empty()
+                    || min_seed_length[m] == this->graph_.get_k()
+                                                - suffix_seeds[m][0].get_offset());
+                min_seed_length[m] = s--;
+                suffix_seeds[m].clear();
+            }
+
+            // e.g., match: $$$ATG, want ATG$$$
```

Comment:
`ATG$$$` doesn't exist (there may be only a single sentinel in the end).
Or you mean a wildcard symbol, not the sentinel? Then write `ATG***`.